### DISCUSSION TITLE
Built-in vs. User defined Functions Performance

### DISCUSSION THREAD

**Discussion by jaslamMiktex** [gap-system/gap#5517] [Ideas]
I would like to report a performance difference of 30X+ factor when using 
   **TransitiveClosureBinaryRelation(rel)**
    vs.
My user-defined function-based version of the transitive closure of a graph having n^3 vertices.  

Here n is the problem size parameter-- vertices in another graph (Graph Isomorphism Problem).
TransitiveClosureBinaryRelation() runs 30X faster than my user-defined GAP program.

I am using the Floyd-Warshall algorithm with O(N^3) time complexity. I do not think anything is missing in my implementation of the F-W algorithm.
So I wonder if there is any better alternative to using TransitiveClosureBinaryRelation(), which was suggested by Max.
PS: Even this test **IsTransitiveBinaryRelation(rel)**  runs 30 times faster on the TransitiveClosureBinaryRelation generated relation.
Thank you

**Comment by james-d-mitchell** [comment]
You could use the Digraphs package which contains equivalent functionality, and some of the routines are implemented in c, including our implementation of the Floyd-Warshall Algorithm, which is probably faster than the library code.

**Comment by fingolfin** [comment]
I am confused about this report. You report that a highly tuned built-in function written by experts is a lot faster than the code you wrote. No offense, but this is not exactly surprising, especially if you are a relative beginner at writing GAP code and don't know about pitfalls etc.

Maybe you are (implicitly) meaning to ask about help for improving your code? Or what is it you are looking to achieve by your report? Perhaps you can clarify that, then you might get help.

Or maybe you really just wanted to report that "GAP is much faster than my own code", but that's not very interesting by itself, I am afraid...

**Reply by james-d-mitchell** [reply]
I somehow read that the performance of the library code was worse than the user code, but on rereading, I agree with @fingolfin, @jaslamMiktex it's not completely clear what you are asking here, perhaps you can clarify?

**Reply by jaslamMiktex** [reply]
Thank you for attending to this.
Yes, my "report" is not particularly interesting or useful. I did mean my code is 30X slower. BTW, I could not  be that bad on GAP. Have had years of development experience in C/C++/Java, etc.
I had assumed that the Digraphs  package would not be any different in performance, and so that is a pending task. Although achieving 30x performance difference is not likely by any means.  (it would be difficult to beat the asymptotic barrier)
Although I am aware of the Permanent() function of GAP which is incredibly fast, up to n =100.

So, the main reason for this "report" was that there is no alternative to TransitiveClosureBinaryRelation() when performance is critical. Max had advised against using it, and which I am still unable to understand why.
Thank you all.
